ya_gpt_bot/services/impl/conversation_service.py

```diff
@@ -0,0 +1,101 @@
+    where
+        chat_id = :chat_id
+),
+windowed as (
+    select
+        ts,
+        sum(length(full_message) + 1) over (order by ts desc) < :context_size as fit_in_context
+    from
+        agg
+)
+select min(ts) as min_ts
+from windowed
+where fit_in_context;
+"""
+)
+
+
+class ConversationService:
+    """Service to get and update conversations."""
+
+    def __init__(self, engine: AsyncEngine):
+        """init"""
+        self.__engine = engine
+
+    async def get_messages_within_context(self, chat_id: int, context_length: int) -> list[str]:
+        """get messages withing defined context from the chat and clear messages that are out of context"""
+        async with self.__engine.connect() as conn:
+            result = (await conn.execute(MIN_TS_SQL, {"context_size": context_length, "chat_id": chat_id})).fetchone()
+            min_ts = result[0].replace(tzinfo=None) if result else None
+
+            if min_ts:
+                full_message_expr = func.concat(
+                    t_conversation.c.user_from,
+                    ",",
+                    func.coalesce(t_conversation.c.user_to, ""),
+                    ",",
+                    t_conversation.c.text,
+                ).label("full_message")
+                select_results = await conn.execute(
+                    select(full_message_expr)
+                    .where(t_conversation.c.message_timestamp >= min_ts, t_conversation.c.chat_id == chat_id)
+                    .order_by(t_conversation.c.message_timestamp)
+                )
+                await conn.execute(
+                    delete(t_conversation).where(
+                        t_conversation.c.message_timestamp < min_ts, t_conversation.c.chat_id == chat_id
+                    )
+                )
+                await conn.commit()
+                return [m[0] for m in select_results.fetchall()]
+
+    async def save_message(  # pylint: disable=too-many-arguments
+        self, chat_id: int, from_name: str, to_name: str, message_timestamp: datetime.datetime, text: str
+    ):
+        """save messages to conversation table"""
+        no_tz = message_timestamp.replace(tzinfo=None)
+        async with self.__engine.connect() as conn:
+            try:
+                await conn.execute(
+                    insert(t_conversation).values(
+                        chat_id=chat_id,
+                        user_from=from_name,
+                        user_to=to_name,
+                        message_timestamp=no_tz,
+                        text=text,
+                    )
+                )
+            except IntegrityError:
+                await conn.rollback()
+                await conn.execute(
+                    insert(t_conversation).values(
+                        chat_id=chat_id,
+                        user_from=from_name,
+                        user_to=to_name,
+                        message_timestamp=no_tz,
+                        text=text,
+                    )
```

Comment:
Could you elaborate on this? What could be the cause of integrity error and why a single retry should fix the problem?

Reply:
It's just copypaste from [messages](https://github.com/kanootoko/ya_gpt_telegram_bot/blob/195a22351b9fe708c04695f76a70a007972305c3/ya_gpt_bot/db/operations/messages.py#L23) db operations 🙂

Reply:
Let it be then, I'll try to refactor both parts later